### DISCUSSION TITLE
[Item/Text] 一个可以关联注释的模板

### DISCUSSION THREAD

**Discussion by mulei000** [windingwind/zotero-better-notes#1101] [Note Templates]
### Description

注释关联模板（无空行），将指定标签的注释转换成笔记（可提取出figure标签中的图片），标签可自行修改，需配合style使用。

参考视频链接：https://www.bilibili.com/video/BV1paWeepEfW/?spm_id_from=333.880.my_history.page.click&vd_source=db6a41b2e759a9514fa2c897b0f7eb2d

模板我已做修改。

### Screenshots

![image](https://github.com/user-attachments/assets/ad4dc8ca-22af-4f50-9a2c-2aca9fac02e3)

### Test on Zotero version

7.0.2 (64-bit)

### Test on Better Notes version

2.0.3

### Template Type

Item

### Template Share Code

```yaml
// @author mulei
${await (async () => {
    const getAnnotationsByTag = async (tagName) => {
        const pdfItem = await topItem.getBestAttachment();
        let result = "";
        for (let annoItem of pdfItem.getAnnotations()) {
            if (annoItem.getTags().find(i => i.tag.includes(tagName))) {
                const res = Zotero.EditorInstanceUtilities.serializeAnnotations(
                    [
                        {
                            ...(await Zotero.Annotations.toJSON(annoItem)), ...{ attachmentItemID: annoItem.parentID }
                        }
                    ]
                );
                result += res.html;
            }
        }
        return result;
    };

    // 辅助函数：检测字符串是否包含中文字符
    const containsChinese = (text) => /[\u4e00-\u9fa5]/.test(text);

    // 提取影响因子
    let impactFactor = Zotero.ZoteroStyle.api.renderCell(topItem, "publicationTags").textContent.trim();
    let impactFactorValues = impactFactor.match(/\d+(\.\d+)?/g); // 提取数字部分

    // 获取翻译标题，如果没有或者为空，则使用原始标题
    let titleTranslationMatch = topItem.getField("extra").match(/titleTranslation:(.+)/);
    let titleTranslation = (titleTranslationMatch && titleTranslationMatch[1].trim()) ? titleTranslationMatch[1].trim() : topItem.getField("title");

    // 检查标题是否为中文
    let isChineseTitle = containsChinese(topItem.getField("title"));

    // 根据标题语言决定IFQ的显示方式
    let impactFactorDisplay = isChineseTitle ? impactFactor : (impactFactorValues ? impactFactorValues.join(' / ') : '');

    // 获取作者显示内容
    let creators = topItem.getCreators();
    let authorDisplay = "";
    if (isChineseTitle) {
        // 中文标题：显示第一作者全名，格式为“姓 前名”
        if (creators.length > 0) {
            let firstAuthor = creators[0];
            let fullName = `${firstAuthor.lastName} ${firstAuthor.firstName}`;
            authorDisplay = fullName;
        }
    } else {
        // 非中文标题：显示第一作者姓氏并附加“et al.”
        if (creators.length > 0) {
            let firstAuthorLastName = creators[0].lastName;
            authorDisplay = `${firstAuthorLastName} et al.`;
        }
    }

    // 获取各部分注释内容
    const 创新点 = await getAnnotationsByTag('创新点');
    const 笔记 = await getAnnotationsByTag('笔记');
    const 背景 = await getAnnotationsByTag('背景');
    const 材料 = await getAnnotationsByTag('材料');
    const 方法 = await getAnnotationsByTag('方法');
    const 结论 = await getAnnotationsByTag('结论');
    const 摘录 = await getAnnotationsByTag('摘录');
    const Figure = await getAnnotationsByTag('Figure');

    // 新增部分：提取 Figure 标签的内容
    const getFigureAnnotations = async (item, tag) => {
        let annots = item.getAnnotations();
        annots = annots.filter((annot) => annot.getTags().map(tagObj => tagObj.tag).includes(tag));
        return await Zotero.BetterNotes.api.convert.annotations2html(annots, { noteItem: targetNoteItem });
    };

    const attachments = Zotero.Items.get(topItem.getAttachments()).filter((i) => i.isPDFAttachment());

    let figureRes = "";
    for (let attachment of attachments) {
        const annots = await getFigureAnnotations(attachment, 'Figure');
        figureRes += annots ? annots : "";
    }

    const itemNotes = []; // 需要获取文献的相关笔记
    const notes = itemNotes.filter(noteItem => noteItem.getTags().map(tagObj => tagObj.tag).includes('Figure'));
    const notesWithTags = `${notes.map((noteItem) => {
        const noteLink = Zotero.BetterNotes.api.convert.note2link(noteItem);
        const noteLine = `<p style="color:red; background-color: #efe3da;">📜 Article Note:  <a href="${noteLink}">${noteItem.getNoteTitle() || noteLink}</a></p>
<p>tags: ${noteItem.getTags().map(tagObj => tagObj.tag).join(', ')}</p>
<p> </p>`;
        return noteLine;
    }).join("\n")}`;

    figureRes += notesWithTags;

    // 生成标题，处理中文和英文情况
    let res = `<h1>🔤${titleTranslation} (<span style="background-color: #B1A5E1">${topItem.getField("year")}</span>, <span style="background-color: #CBE1A5">${authorDisplay}</span>) ${topItem.getField("journalAbbreviation")} (${impactFactorDisplay})</h1>
<p><strong><span style="color: #283593">原名：</span></strong>${topItem.getField('title')}</p>
<p><strong><span style="color: #283593">译名：</span></strong>${titleTranslation}</p>
<p><strong><span style="color: #283593">作者：</span></strong>${authorDisplay}</p>
<p><strong><span style="color: #283593">期刊：</span></strong>${topItem.getField('publicationTitle')}</p>
<p><strong><span style="color: #283593">IFQ：</span></strong><span style="">

${{
let space = " ㅤㅤ ㅤㅤ"
return Array.prototype.map.call(
	Zotero.ZoteroStyle.api.renderCell(topItem, "publicationTags").childNodes,
	e => {
		e.innerText =  space + e.innerText + space;
		return e.outerHTML
	}
	).join(space)
}}$
</p >

${(() => {
        const doi = topItem.getField("DOI");
        if (doi) {
            return `<b><span style="color: #283593">DOI：</span></strong> </b><a href="https://doi.org/${topItem.getField('DOI')}">${topItem.getField('DOI')}</a>`;
        } else {
            return `<b><span style="color: #283593">URL：</span></strong> </b><a href="${topItem.getField('url')}">${topItem.getField('url')}</a>`;
        }
    })()}
<p><strong><span style="color: #283593">发表时间：</span></strong>${topItem.getField('date')}</p>

<!-- 本地链接 -->
<tr>
  <td style="color:#193c47; background-color:#dbeedd; padding:8px;">
    ${(() => {
        const attachments = Zotero.Items.get(topItem.getAttachments());
        if (attachments && attachments.length > 0) {
            return `<b><span style="color: #283593">本地链接:</span> </b><a href="zotero://open-pdf/0_${attachments[0].key}">${attachments[0].getFilename()}</a>`;
        } else {
            return `<b><span style="color: #283593">本地链接:</span> </b>`;
        }
    })()}
  </td>
</tr>
<p>
<!-- 摘要 -->
<tr>
  <td style="color:#193c47; background-color:#dbeedd; padding:8px;">
    ${(() => {
        const abstractTranslation = topItem.getField('abstractTranslation');
        if (abstractTranslation) {
            return `<b><span style="color: #283593">摘要翻译：</span> </b><i>${abstractTranslation}</i>`;
        } else {
            return `<b><span style="color: #283593">摘要：</span> </b><i>${topItem.getField('abstractNote')}</i>`;
        }
    })()}
  </td>
</tr>

<!-- 生成标签笔记 -->
${创新点 ? `<h3><span style="color: #A61B41">💡创新点：</span></h3>${创新点}` : ''}

${笔记 ? `<h3><span style="color: #A61B41">♥️笔记：</span></h3>${笔记}` : ''}

${背景 ? `<h3><span style="color: #A61B41">🌏背景：</span></h3>${背景}` : ''}

${材料 ? `<h3><span style="color: #A61B41">💧材料：</span></h3>${材料}` : ''}

${方法 ? `<h3><span style="color: #A61B41">🔬方法：</span></h3>${方法}` : ''}

${结论 ? `<h3><span style="color: #A61B41">📜结论：</span></h3>${结论}` : ''}

${摘录 ? `<h3><span style="color: #A61B41">🚩摘录：</span></h3>${摘录}` : ''}

${figureRes ? `<h3><span style="color: #A61B41">🖼️Figure 笔记：</span></h3>${figureRes}` : ''}

`;

    return res;
})()}

**Comment by jyhuang81** [comment]
显示不了图片？

**Reply by mulei000** [reply]
可以显示图片，给图片加figure的标签

**Reply by jyhuang81** [reply]
粘贴，新建模板后显示“Preview not available for template ”。从模板建立笔记，内容也是空白的。从模板新建笔记没有任何反应，预览也没显示。
Zotero 7.0.3-beta.2+10d27b8f6 (64-bit)
Better Notes for Zotero 2.0.3

**Reply by mulei000** [reply]
你看的中文文献吧，我根据这个up主的新内容又重新修改了中英文通用的模板，要注意和视频里面提到的和style、tag-action的联用步骤，目前硕博论文中也不可用

**Reply by wzr-ms** [reply]
> 可以显示图片，给图片加figure的标签

请问如何添加figure标签呀？

**Reply by DWCY3574** [reply]
> 可以显示图片，给图片加数字的标签

最新版的软件和插件，加了figure标签也不会显示图片

**Reply by wzr-ms** [reply]
可以显示啊，你是不是标签的string没改对啊？你再检查一下

**Reply by DWCY3574** [reply]
其他的可以显示，figure标签的string还要改吗

**Reply by DWCY3574** [reply]
> 可以显示啊，你是不是标签的string没改对啊？你再检查一下

大神教教我怎么改

**Reply by DWCY3574** [reply]
![image](https://github.com/user-attachments/assets/b5821dae-663d-479c-9e31-9ba9b11243c0)
这个大小写有要求吗，style中颜色标签为小写的figure

> 可以显示啊，你是不是标签的string没改对啊？你再检查一下

**Reply by wzr-ms** [reply]
好像不用改就那样也可以，你得用那个图片框选的按钮，你得笔记模板有那个tag就行

**Reply by wzr-ms** [reply]
改成小写的变量

**Reply by wzr-ms** [reply]
最好别用大写

**Reply by DWCY3574** [reply]
试过了小写的也不行

**Reply by DWCY3574** [reply]
你是怎么改的呢

> 最好别用大写

**Comment by ZXM614** [comment]
作者大大，想请问预览正常，用的时候是空白的，这是为什么？
![image](https://github.com/user-attachments/assets/156b1673-5cdc-4561-a8ce-0ffc4cc95ec7)

**Comment by ZXM614** [comment]
![Uploading image.png…]()

**Comment by mulei000** [comment]
母鸡，我自己使用时正常，可能是你改标签的时候出问题了



发自我的iPhone


------------------ 原始邮件 ------------------
发件人: ZXM614 ***@***.***&gt;
发送时间: 2024年9月14日 20:07
收件人: windingwind/zotero-better-notes ***@***.***&gt;
抄送: mulei000 ***@***.***&gt;, Author ***@***.***&gt;
主题: Re: [windingwind/zotero-better-notes] [Item/Text] 一个可以关联注释的模板 (Discussion #1101)

**Comment by ZXM614** [comment]
我还没改标签，只复制过来就不行啦



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2024年9月14日(周六) 晚上8:08
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [windingwind/zotero-better-notes] [Item/Text] 一个可以关联注释的模板 (Discussion #1101)




 母鸡，我自己使用时正常，可能是你改标签的时候出问题了 
 
 
 
 发自我的iPhone 
 
 
 ------------------ 原始邮件 ------------------ 
 发件人: ZXM614 ***@***.***&amp;gt; 
 发送时间: 2024年9月14日 20:07 
 收件人: windingwind/zotero-better-notes ***@***.***&amp;gt; 
 抄送: mulei000 ***@***.***&amp;gt;, Author ***@***.***&amp;gt; 
 主题: Re: [windingwind/zotero-better-notes] [Item/Text] 一个可以关联注释的模板 (Discussion #1101)

**Comment by mulei000** [comment]
判断不了，模版报错不会提醒问题，我自己测试师妹问题的



发自我的iPhone


------------------ 原始邮件 ------------------
发件人: ZXM614 ***@***.***&gt;
发送时间: 2024年9月14日 20:33
收件人: windingwind/zotero-better-notes ***@***.***&gt;
抄送: mulei000 ***@***.***&gt;, Author ***@***.***&gt;
主题: Re: [windingwind/zotero-better-notes] [Item/Text] 一个可以关联注释的模板 (Discussion #1101)





 我还没改标签，只复制过来就不行啦 
 
 
 
 ---原始邮件--- 
 发件人: ***@***.***&amp;gt; 
 发送时间: 2024年9月14日(周六) 晚上8:08 
 收件人: ***@***.***&amp;gt;; 
 抄送: ***@***.******@***.***&amp;gt;; 
 主题: Re: [windingwind/zotero-better-notes] [Item/Text] 一个可以关联注释的模板 (Discussion #1101) 
 
 
 
 
  母鸡，我自己使用时正常，可能是你改标签的时候出问题了  
   
   
   
  发自我的iPhone  
   
   
  ------------------ 原始邮件 ------------------  
  发件人: ZXM614 ***@***.***&amp;amp;gt;  
  发送时间: 2024年9月14日 20:07  
  收件人: windingwind/zotero-better-notes ***@***.***&amp;amp;gt;  
  抄送: mulei000 ***@***.***&amp;amp;gt;, Author ***@***.***&amp;amp;gt;  
  主题: Re: [windingwind/zotero-better-notes] [Item/Text] 一个可以关联注释的模板 (Discussion #1101)

**Comment by ZXM614** [comment]
好的，感谢您的回复



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2024年9月14日(周六) 晚上8:58
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [windingwind/zotero-better-notes] [Item/Text] 一个可以关联注释的模板 (Discussion #1101)




 判断不了，模版报错不会提醒问题，我自己测试师妹问题的 
 
 
 
 发自我的iPhone 
 
 
 ------------------ 原始邮件 ------------------ 
 发件人: ZXM614 ***@***.***&amp;gt; 
 发送时间: 2024年9月14日 20:33 
 收件人: windingwind/zotero-better-notes ***@***.***&amp;gt; 
 抄送: mulei000 ***@***.***&amp;gt;, Author ***@***.***&amp;gt; 
 主题: Re: [windingwind/zotero-better-notes] [Item/Text] 一个可以关联注释的模板 (Discussion #1101) 
 
 
 
 
 
  我还没改标签，只复制过来就不行啦  
   
   
   
  ---原始邮件---  
  发件人: ***@***.***&amp;amp;gt;  
  发送时间: 2024年9月14日(周六) 晚上8:08  
  收件人: ***@***.***&amp;amp;gt;;  
  抄送: ***@***.******@***.***&amp;amp;gt;;  
  主题: Re: [windingwind/zotero-better-notes] [Item/Text] 一个可以关联注释的模板 (Discussion #1101)  
   
   
   
   
   母鸡，我自己使用时正常，可能是你改标签的时候出问题了   
     
     
     
   发自我的iPhone   
     
     
   ------------------ 原始邮件 ------------------   
   发件人: ZXM614 ***@***.***&amp;amp;amp;gt;   
   发送时间: 2024年9月14日 20:07   
   收件人: windingwind/zotero-better-notes ***@***.***&amp;amp;amp;gt;   
   抄送: mulei000 ***@***.***&amp;amp;amp;gt;, Author ***@***.***&amp;amp;amp;gt;   
   主题: Re: [windingwind/zotero-better-notes] [Item/Text] 一个可以关联注释的模板 (Discussion #1101)

**Comment by mulei000** [comment]
通过style设置一个颜色为figure，再用action写一个脚本将对应颜色设置为不同标签。我发的视频的链接里面有方法，我只是改了模版，方法与这个up主相同



发自我的iPhone


------------------ 原始邮件 ------------------
发件人: Wang Zhangren ***@***.***&gt;
发送时间: 2024年9月17日 15:04
收件人: windingwind/zotero-better-notes ***@***.***&gt;
抄送: mulei000 ***@***.***&gt;, Author ***@***.***&gt;
主题: Re: [windingwind/zotero-better-notes] [Item/Text] 一个可以关联注释的模板 (Discussion #1101)

**Reply by wzr-ms** [reply]
这个能不能和一个图片突出的插件实现picture2那个实现一种新的联动

**Comment by mulei000** [comment]
手动可以，你自己提取完图片以后，去全选后加上figure标签（确认一下你填写的标签的大小写
）

发自我的iPhone


------------------ 原始邮件 ------------------
发件人: Wang Zhangren ***@***.***&gt;
发送时间: 2024年9月19日 16:09
收件人: windingwind/zotero-better-notes ***@***.***&gt;
抄送: mulei000 ***@***.***&gt;, Author ***@***.***&gt;
主题: Re: [windingwind/zotero-better-notes] [Item/Text] 一个可以关联注释的模板 (Discussion #1101)





 
这个能不能和一个图片突出的插件实现picture2那个实现一种新的联动

**Comment by wzr-ms** [comment]
谢谢友友，他对那个标签的命名采用Figure1 Figure2在想怎么改改就能避免手动哈哈哈


发件人：mulei000 ***@***.***>
发送日期：2024-09-19 16:14:44
收件人：windingwind/zotero-better-notes ***@***.***>
抄送人："Wang Zhangren" ***@***.***>, Comment ***@***.***>
主 题：Re: [windingwind/zotero-better-notes] [Item/Text] 一个可以关联注释的模板 (Discussion #1101)


手动可以，你自己提取完图片以后，去全选后加上figure标签（确认一下你填写的标签的大小写
）

发自我的iPhone


------------------ 原始邮件 ------------------
发件人: Wang Zhangren ***@***.***&gt;
发送时间: 2024年9月19日 16:09
收件人: windingwind/zotero-better-notes ***@***.***&gt;
抄送: mulei000 ***@***.***&gt;, Author ***@***.***&gt;
主题: Re: [windingwind/zotero-better-notes] [Item/Text] 一个可以关联注释的模板 (Discussion #1101)






这个能不能和一个图片突出的插件实现picture2那个实现一种新的联动

**Comment by mulei000** [comment]
使用action 脚本


/**
&nbsp;* 直接添加"Figure"标签
&nbsp;* @author Mulei000
&nbsp;*/


// 直接添加"Figure"标签
item.addTag("Figure");
return "Tag added: Figure";


发自我的iPhone


------------------ 原始邮件 ------------------
发件人: Wang Zhangren ***@***.***&gt;
发送时间: 2024年9月19日 18:33
收件人: windingwind/zotero-better-notes ***@***.***&gt;
抄送: mulei000 ***@***.***&gt;, Author ***@***.***&gt;
主题: Re: [windingwind/zotero-better-notes] [Item/Text] 一个可以关联注释的模板 (Discussion #1101)





 谢谢友友，他对那个标签的命名采用Figure1 Figure2在想怎么改改就能避免手动哈哈哈
 
 
 发件人：mulei000 ***@***.***&gt;
 发送日期：2024-09-19 16:14:44
 收件人：windingwind/zotero-better-notes ***@***.***&gt;
 抄送人："Wang Zhangren" ***@***.***&gt;, Comment ***@***.***&gt;
 主 题：Re: [windingwind/zotero-better-notes] [Item/Text] 一个可以关联注释的模板 (Discussion #1101)
 
 
 手动可以，你自己提取完图片以后，去全选后加上figure标签（确认一下你填写的标签的大小写
 ）
 
 发自我的iPhone
 
 
 ------------------ 原始邮件 ------------------
 发件人: Wang Zhangren ***@***.***&amp;gt;
 发送时间: 2024年9月19日 16:09
 收件人: windingwind/zotero-better-notes ***@***.***&amp;gt;
 抄送: mulei000 ***@***.***&amp;gt;, Author ***@***.***&amp;gt;
 主题: Re: [windingwind/zotero-better-notes] [Item/Text] 一个可以关联注释的模板 (Discussion #1101)
 
 
 
 
 
 
 这个能不能和一个图片突出的插件实现picture2那个实现一种新的联动

**Comment by iiang22** [comment]
请问怎么可以显示全部作者？代码需要怎么修改？不知道哪里可以学习相关的代码知识 呜呜

**Reply by wzr-ms** [reply]
为啥一定要显示全部作者，这个你可以查查现有的帖子，我没这个需求就没弄

**Reply by windingwind** [reply]
<img width="554" alt="image" src="https://github.com/user-attachments/assets/2f72a1a7-3573-40ae-9ac4-e1fc6a03be06">

**Reply by wzr-ms** [reply]
友友觉得这种关于带有文字的OCR怎么解决，主要还是成本问题，API这些，每个月花费表感觉不少，我是学生哈哈哈

**Comment by iiang22** [comment]
非常感谢！
&nbsp;


刘思洋
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2024年9月27日(星期五) 下午4:33
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [windingwind/zotero-better-notes] [Item/Text] 一个可以关联注释的模板 (Discussion #1101)














Image.png（在网络上查看）
-
直接回复这封邮件，在GitHub上查看，或取消订阅.
***@***.*** &gt;
